AutoCollection/CorrelationContextManager.ts

```diff
@@ -0,0 +1,152 @@
+        }
+        return Zone.current.get("context");
+    }
+
+    /**
+     *  A helper to generate objects conforming to the CorrelationContext interface
+     */
+    public static generateContextObject(parentId?: string, operationName?: string, operationId?: string): CorrelationContext {
+        operationId = operationId || Util.newGuid();
+        parentId = parentId || operationId;
+        
+        if (this.enabled) {
+            return {
+                operation: {
+                    name: operationName,
+                    id: operationId,
+                    parentId: parentId
+                },
+                customProperties: {}
+            };
+        }
+
+        return null;
+    }
+
+    /** 
+     *  Runs a function inside a given Context.
+     *  All logical children of the execution path that entered this Context
+     *  will receive this Context object on calls to GetCurrentContext.
+     */
+    public static runWithContext(context: CorrelationContext, fn: ()=>any) {
+        if (CorrelationContextManager.enabled) {
+            var newZone = Zone.current.fork({
+                name: "AI-" + ((context && context.operation.parentId) || "Unknown"), 
+                properties: {context: context}
+            });
+            newZone.run(fn);
+        } else {
+            fn();
+        }
+    }
+
+    /** 
+     *  Patches a callback to restore the correct Context when getCurrentContext 
+     *  is run within it. This is necessary if automatic correlation fails to work
+     *  with user-included libraries.
+     * 
+     *  The supplied callback will be given the same context that was present for 
+     *  the call to wrapCallback.  */
+    public static wrapCallback<T extends Function>(fn: T): T {
+        if (CorrelationContextManager.enabled) {
+            return Zone.current.wrap(fn, "User-wrapped method");
+        }
+        return fn;
+    }
+
+    /**
+     *  Enables the CorrelationContextManager.
+     */
+    public static enable() {
+        if (!this.isNodeVersionCompatible()) {
+            this.enabled = false;
+            return;
+        }
+
+        // Run patches first
+        if (!this.hasEverEnabled) {
+            this.hasEverEnabled = true;
+            this.patchRedis();
+        }
+
+        this.enabled = true;
+    }
+
+    /**
+     *  Disables the CorrelationContextManager.
+     */
+    public static disable() {
+        this.enabled = false;
+    }
+
+
+    /**
+     *  Reports if the CorrelationContextManager is able to run in this environment
+     */
+    public static isNodeVersionCompatible() {
+        return typeof Zone !== 'undefined'
+    }
+
+    // Patch methods that manually go async that Zone doesn't catch
+    private static requireForPatch(module: string) {
+        var req = null;
+        try {
+            req = require(module);
+        } catch (e) {
+            return null;
+        }
+        return req;
+    }
+
+    private static patchRedis() {
```

Comment:
write comment on what you are doing here. Also it would be great to share the docs (like `README.md` or API reference) on how other third party libraries can be patched if needed.